### DISCUSSION TITLE
Add an extension that provides schema validation and auto-generated API documentation

### DISCUSSION THREAD

**Discussion by lewoudar** [miguelgrinberg/microdot#190] [Ideas]
Hi everyone,
First of all, I want to thank the author for this project. I think I will try it when I have some free time. It can be really useful for quick demos or tutorials in my case.
One thing I think could make the project even more interesting is to have an extension like [Quart-schema](https://pgjones.gitlab.io/quart-schema/) but for Microdot. Since building APIs is common these days, it can help drive more users to this project.

If you find the idea interesting, you may consider using [msgspec](https://jcristharif.com/msgspec/) instead of Pydantic, which is smaller and faster 🙂

**Comment by miguelgrinberg** [comment]
I agree in principle, but the fact that there are several options without a clear leader for CPython (Pydantic, Marshmallow, WTForms, msgspec) and absolutely none that I know for MicroPython makes it harder for me to decide what the best option is. So right now I don't think I am ready to add an official extension, though I will likely do something once I have more clarity.

**Comment by lewoudar** [comment]
I see your point. But IMHO, WTForms and Marshmallow are a bit old now. Pydantic and msgspec are more natural with more and more users using typing in their projects. Pydantic is quite popular nowadays if we look at the number of downloads on [PyPI](https://www.pepy.tech/).
msgspec is also fine, smaller, and faster than the other options.

**Reply by miguelgrinberg** [reply]
Well, I use Marshmallow, so that's that. Just proves my point that no matter the choice, there's always going to be people who'll disagree. I try to be as unopinionated as I can, so I do not want to make a choice, since there are many reasonable opinions just like yours or mine that make the case for X being the best one of all, regardless of X.

**Reply by lewoudar** [reply]
Just to be clear, I have never said that I think Pydantic is the best (although I use it a lot to be honest), but to say that there is no clear leader is not entirely true, which is why I point you to a website where you can check download traffic. Msgspec is probably a better option for MicroPython if space is a real issue.

**Reply by miguelgrinberg** [reply]
MicroPython does not have any compatible validation libraries. Do you know for a fact that msgpack is compatible or are just guessing? A package designed for CPython is unlikely to work.

**Reply by lewoudar** [reply]
oops, I was just guessing, sorry.
